pulsar-jms-tracing/src/main/java/com/datastax/oss/pulsar/jms/tracing/BrokerTracing.java

```diff
@@ -0,0 +1,484 @@
+import org.apache.pulsar.common.api.proto.MessageMetadata;
+import org.apache.pulsar.common.intercept.InterceptException;
+
+@Slf4j
+public class BrokerTracing implements BrokerInterceptor {
+
+  public enum EventReasons {
+    ADMINISTRATIVE,
+    MESSAGE,
+    TRANSACTION,
+    SERVLET,
+  }
+
+  private static final TraceLevel defaultTraceLevel = TraceLevel.BASIC;
+
+  private static final LoadingCache<PulsarService, Set<EventReasons>> jmsTracingEventList =
+      CacheBuilder.newBuilder()
+          .maximumSize(10)
+          .expireAfterWrite(1, TimeUnit.MINUTES)
+          .build(
+              new CacheLoader<PulsarService, Set<EventReasons>>() {
+                public Set<EventReasons> load(PulsarService pulsar) {
+                  // dynamic config can get reloaded without service restart
+                  String events =
+                      pulsar
+                          .getConfiguration()
+                          .getProperties()
+                          .getProperty("jmsTracingEventList", "");
+                  log.debug("read jmsTracingEventList: {}", events);
+
+                  Set<EventReasons> enabledEvents = new HashSet<>();
+
+                  for (String event : events.split(",")) {
+                    try {
+                      enabledEvents.add(EventReasons.valueOf(event.trim()));
+                    } catch (IllegalArgumentException e) {
+                      log.warn("Invalid event: {}. Skipping", event);
+                    }
+                  }
+
+                  return enabledEvents;
+                }
+              });
+
+  private static final LoadingCache<PulsarService, TraceLevel> traceLevelForService =
+      CacheBuilder.newBuilder()
+          .maximumSize(10)
+          .expireAfterWrite(1, TimeUnit.MINUTES)
+          .build(
+              new CacheLoader<PulsarService, TraceLevel>() {
+                public TraceLevel load(PulsarService pulsar) {
+                  String level =
+                      pulsar
+                          .getConfiguration()
+                          .getProperties()
+                          .getProperty("jmsTracingLevel", defaultTraceLevel.toString());
+                  try {
+                    return TraceLevel.valueOf(level);
+                  } catch (IllegalArgumentException e) {
+                    log.warn(
+                        "Invalid tracing level: {}. Using default: {}", level, defaultTraceLevel);
+                    return defaultTraceLevel;
+                  }
+                }
+              });
+
+  private static final LoadingCache<Subscription, TraceLevel> traceLevelForSubscription =
+      CacheBuilder.newBuilder()
+          .expireAfterWrite(10, TimeUnit.SECONDS)
+          .build(
+              new CacheLoader<Subscription, TraceLevel>() {
+                public TraceLevel load(Subscription sub) {
+                  if (sub.getSubscriptionProperties() == null
```

Comment:
nit:
I suggest to put sub.getSubscriptionProperties() in a variable.
The map is always unmodifiable but it can be changed, so we could have some subtle bug here